src/modelfree/common/utils.py

```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
+
+                ep_ret = sum(agent_dicts[env_idx]['rewards'])
+                self.full_traj_dicts[agent_idx]['episode_returns'].append(np.array([ep_ret]))
+
+                for key in data_keys:
+                    # consolidate episode data and append to long-term data dict
+                    episode_key_data = np.array(agent_dicts[env_idx][key])
+                    self.full_traj_dicts[agent_idx][key].append(episode_key_data)
+                agent_dicts[env_idx] = defaultdict(list)
+                # so as not to double-count number of episodes
+                self.num_completed += int(agent_idx == 0)
+                if self.num_completed >= self.num_traj_to_save:
+                    self.save_traj(agent_idx)
+
+    def save_traj(self, agent_idx):
+        if self.already_saved[agent_idx]:
+            print(f'Warning: already saved trajectories for agent {agent_idx}')
+        dump_dict = {
+            k: np.concatenate(v, axis=0)
+            for k, v in self.full_traj_dicts[agent_idx].items()}
+        save_path = os.path.join(self.traj_dir, f'agent_{agent_idx}.npz')
+        np.savez(save_path, **dump_dict)
+        self.already_saved[agent_idx] = True
+
+
+def simulate(venv, policies, render=False, record_traj=False, traj_dir='data/',
+             num_traj_to_save=None):
     """
     Run Environment env with the agents in agents
     :param venv(VecEnv): vector environment.
     :param policies(list<BaseModel>): a policy per agent.
     :param render: true if the run should be rendered to the screen
+    :param record_traj: whether to save trajectory data in stable_baselines.GAIL format
+    :param traj_dir: where to save the trajectories
+    :param num_traj_to_save: when to save trajectories since this function is a generator
     :return: streams information about the simulation
     """
     observations = venv.reset()
     dones = [False] * venv.num_envs
     states = [None for _ in policies]
 
+    if record_traj:
+        if num_traj_to_save is None:
+            raise ValueError("Must set number of trajectories to save in order to save them.")
+        recorder = TrajectoryRecorder(venv.num_envs, len(policies), num_traj_to_save, traj_dir)
+
     while True:
         if render:
             venv.render()
 
         actions = []
         new_states = []
-        for policy, obs, state in zip(policies, observations, states):
+        prev_observations = observations
+        for idx, (policy, obs, state) in enumerate(zip(policies, observations, states)):
             act, new_state = policy.predict(obs, state=state, mask=dones)
             actions.append(act)
             new_states.append(new_state)
         actions = tuple(actions)
         states = new_states
 
         observations, rewards, dones, infos = venv.step(actions)
+        if record_traj:
+            recorder.record_traj(rewards, actions, dones, prev_observations)
```

Comment:
Nitpick: I'd find `prev_observations, actions, rewards, dones` a more intuitive order. We see `prev_observations` first; we choose `actions`; then we receive `rewards` and `dones`.